### PR TITLE
OAP-stream integration with Banyandb

### LINKED ISSUE #7732: [Core] Stream data processing
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In https://github.com/apache/skywalking/issues/6922, we accomplished most features of the trace module and other core components that support processing tracing data. We even did a banyandb java client for OAP. 

Unfortunately, when I was implementing the index module, there was a defect in the original design that blocked us to make the trace feature happen.  Based on that, I set up this new issue to track current tasks.

Now we selected `Stream` as the official notion to ingest trace and log items. The `trace` module will be dropped once this transferring get done.

- [x] Upgrade API to v2. In v2, I introduce `stream` to substitute `trace` and tweak index relevant objects.
- [x] Stream module.
- [x] TSDB module. It's a time-series database engine to handle all of the data models.
- [x] Several index packages. Including inverted index, tree(LSM-tree) index, term dictionary, and query execution tree.
- [x] Prune outdated modules that are `trace`, `database`  and `index`.
- [x] Query module. Adjust logical execution plan accordingly. (@lujiajing1126)
- [x] Liaison. Adapt to the new stream protocols.
- [x] Metadata empowered by embed Etcd. (@lujiajing1126) 
- [x] Java client. Adapt to the new stream protocols. (@lujiajing1126)
- [x] OAP storage layer https://github.com/apache/skywalking/issues/7412.(@lujiajing1126)

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lujiajing1126**
### Introduce OAP integration with BanyanDB stream 
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #7732
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

This PR intends to integrate stream related entities to BanyanDB. And this is the last task listed in issue #7732 

I will open another issue to track the remaining features which are not currently supported by BanyanDB but are used in the query protocol of OAP server.

UT and documents will be added later.

@hanahmily

### CODE REVIEW

**Comment by wu-sheng** [comment]
I think you may use a wrong milestone? That milestone should track issues for server side release.

**Comment by wu-sheng** [comment]
Could you share what is the pattern of `AbstractBanyanDBDeserializer` about? I am sure this pattern is not being used in other storage implementation.

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/AbstractBanyanDBDAO.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    protected AbstractBanyanDBDAO(BanyanDBStorageClient client) {
+        super(client);
+    }
+
+    protected <T> List<T> query(Class<T> clazz, QueryBuilder builder) {
+        return this.query(clazz, builder, 0, 0);
+    }
+
+    protected <T> List<T> query(Class<T> clazz, QueryBuilder builder, long startTimestamp, long endTimestamp) {
+        AbstractBanyanDBDeserializer<T> deserializer = BanyanDBDeserializerFactory.INSTANCE.findDeserializer(clazz);
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/AbstractBanyanDBDAO.java:41]: This kind of factory seems less performance? Most DAO implementations are simple logic, could we make BanyanDB DAOs more consistent with others?

Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/AbstractBanyanDBDAO.java:41]: The reason behinds the codes are, this kind of logic could be optimized and very custom for every single cases. Adding a deserializer seems moving logic of one method to other places, and only makes things more complex.

Comment (lujiajing1126) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/AbstractBanyanDBDAO.java:41]: Make sense to me. I could move the deserializer to each query DAO to optimize this hashmap search away in the sense that almost each kind of queryDAO uses its own deserializer.

**Comment by lujiajing1126** [comment]
> Could you share what is the pattern of `AbstractBanyanDBDeserializer` about? I am sure this pattern is not being used in other storage implementation.

The reason to introduce the `AbstractBanyanDBDeserializer` is that the storage build for BanyanDB is asymmetric, i.e. what we write is several lists of `SerializableTag`s while we read `TagAndValue` from database. Both of these lists are strictly ordered which should comply with the order of schema definition and the order of projection declaration.

Another motivation is that default implementation of `storageBuilder` within the DAO is less performant since binary data is always encoded to the base64 format which is not necessary in the case of BanyanDB. We natively support binary data format.

Besides, I suppose it is not necessary to fetch the full storage entity as an intermediate representation while querying. We could do the projection to reduce bandwidth usage.

**Comment by wu-sheng** [comment]
Writing as async is the common case for all storage implementation. I am not sure what do you mean. All metrics/logs/traces in all storage implementations are async. Please check the exiting codes.

**Comment by wu-sheng** [comment]
> Besides, I suppose it is not necessary to fetch the full storage entity as an intermediate representation while querying. We could do the projection to reduce bandwidth usage.

This is also unnecessary for query. We don't do that in other storage implementation too.

**Comment by wu-sheng** [comment]
> Another motivation is that default implementation of storageBuilder within the DAO is less performant since binary data is always encoded to the base64 format which is not necessary in the case of BanyanDB. We natively support binary data format.

StorageBuilderFactory and StorageBuilder make sense to me to get rid of Map-based mechanism. My question is more about deserializer in query stage.

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        } catch (NoSuchMethodException noSuchMethodException) {
+            log.error("cannot find method storage2Entity", noSuchMethodException);
+            throw new RuntimeException("cannot find method storage2Entity");
+        }
+    }
+
+    @Override
+    public INoneStreamDAO newNoneStreamDao(StorageBuilder storageBuilder) {
+        try {
+            Class<?> returnType = storageBuilder.getClass().getMethod("storage2Entity", Map.class).getReturnType();
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java:114]: Why do we need reflect in this kind of DAO?

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            } else if (InstanceTraffic.class.equals(returnType)) {
+                return new BanyanDBMetricsDAO<>(new Metadata.InstanceTrafficBuilder());
+            } else if (EndpointTraffic.class.equals(returnType)) {
+                return new BanyanDBMetricsDAO<>(new Metadata.EndpointTrafficBuilder());
+            } else if (NetworkAddressAlias.class.equals(returnType)) {
+                return new BanyanDBMetricsDAO<>(new NetworkAddressAliasBuilder());
+            } else {
+                throw new IllegalStateException("record type is not supported");
+            }
+        } catch (NoSuchMethodException noSuchMethodException) {
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java:80]: MetricsDao with metadata only? Are you sure this is the right way to do this?
`StorageBuilderFactory#Default` bases on the class scan mechanism. If you want your own, you should create another way, such as codes or annotation to declare this. But never do this kind of reflect and `if/else`

Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java:80]:
```java
@Stream(name = ServiceTraffic.INDEX_NAME, scopeId = DefaultScopeDefine.SERVICE,
    builder = ServiceTraffic.Builder.class, processor = MetricsStreamProcessor.class)
@MetricsExtension(supportDownSampling = false, supportUpdate = false)
@EqualsAndHashCode(of = {
    "name",
    "nodeType"
})
public class ServiceTraffic extends Metrics {
```

This is how Builder is declared and mapped to the entity. I think you should have a class or a class with annotations to declare the override? Such as(only an example)
```java
@StreamBuilderOverride(name = ServiceTraffic.INDEX_NAME,  builder = BanyanDBServiceTrafficBuilder.class)
@StreamBuilderOverride(name = InstanceTraffic.INDEX_NAME,  builder = BanyanDBInstanceTrafficBuilder.class)
public class BanyanDBStorageBuilderFactory
```

File: oap-server/server-storage-plugin/storage-banyandb-plugin/pom.xml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        <groupId>org.apache.skywalking</groupId>
+        <version>8.10.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>storage-banyandb-plugin</artifactId>
+    <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/pom.xml:32]: @kezhenxu94 @hanahmily We should take care of these CVEs by bumping up guava, jetty, etc.

**Comment by lujiajing1126** [comment]
> > Besides, I suppose it is not necessary to fetch the full storage entity as an intermediate representation while querying. We could do the projection to reduce bandwidth usage.
> 
> This is also unnecessary for query. We don't do that in other storage implementation too.

I suppose at least IOTDB implementation is doing this?

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java (hunk 3 above)
Comment (lujiajing1126) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/stream/BanyanDBStorageDAO.java:80]: I missed this part. I will check it.

**Comment by lujiajing1126** [comment]
> > Another motivation is that default implementation of storageBuilder within the DAO is less performant since binary data is always encoded to the base64 format which is not necessary in the case of BanyanDB. We natively support binary data format.
> 
> StorageBuilderFactory and StorageBuilder make sense to me to get rid of Map-based mechanism. My question is more about deserializer in query stage.

But the Map-based `storage2Entity`s are also doing base64 decoding, for example builders in `AlarmRecord`, `BrowserErrorLogRecord` etc.

Besides,

**Comment by wu-sheng** [comment]
> But the Map-based storage2Entitys are also doing base64 decoding, for example builders in AlarmRecord, BrowserErrorLogRecord etc.

Only builder did BASE64, I think we are clear builder should be override.

> I suppose at least IOTDB implementation is doing this?

This is the newest implementation. Why do you learn from it? BanyanDB DAO would be closest to core and native concept. You should not learn from IoTDB implementation.

**Comment by lujiajing1126** [comment]
> > But the Map-based storage2Entitys are also doing base64 decoding, for example builders in AlarmRecord, BrowserErrorLogRecord etc.
> 
> Only builder did BASE64, I think we are clear builder should be override.
> 
> > I suppose at least IOTDB implementation is doing this?
> 
> This is the newest implementation. Why do you learn from it? BanyanDB DAO would be closest to core and native concept. You should not learn from IoTDB implementation.

Well. It has been merged. So I guess it could be a referential impl.

I've checked JDBC impl again. I could understand what you meant. I could move the deserialization logic to the query DAO just what we've done in JDBC, influxdb impl. Then we don't need to the deser abstraction.

**Comment by wu-sheng** [comment]
> I've checked JDBC impl again. I could understand what you meant. I could move the deserialization logic to the query DAO just as what we've done in JDBC, influxdb impl. Then we don't need to the deser abstraction.

Most importantly, let's make Builder and BuilderFactory work to override the core. That is the most important.

**Comment by lujiajing1126** [comment]
> > I've checked JDBC impl again. I could understand what you meant. I could move the deserialization logic to the query DAO just as what we've done in JDBC, influxdb impl. Then we don't need to the deser abstraction.
> 
> Most importantly, let's make Builder and BuilderFactory work to override the core. That is the most important.

What about registering customized `StorageBuilderFactory` which can be done in the `prepare` stage of `ModuleProvider`.

By default, `new StorageBuilderFactory.Default()` is used.

**Comment by wu-sheng** [comment]
> What about registering customized StorageBuilderFactory which can be done in the prepare stage of ModuleProvider.
By default, new StorageBuilderFactory.Default() is used.

Yes, we should override it through this way. Notice, this was designed and added by myself in the very beginning, but we never used in upstream in this way. Please let me know if this mechanism has an issue.

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/deserializer/SegmentRecordDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        record.setSegmentId(row.getId());
+        record.setTraceId((String) searchable.get(0).getValue());
+        record.setIsError(((Number) searchable.get(1).getValue()).intValue());
+        record.setServiceId((String) searchable.get(2).getValue());
+        record.setServiceInstanceId((String) searchable.get(3).getValue());
+        record.setEndpointId((String) searchable.get(4).getValue());
+        record.setLatency(((Number) searchable.get(5).getValue()).intValue());
+        record.setStartTime(((Number) searchable.get(6).getValue()).longValue());
+        final List<TagAndValue<?>> data = row.getTagFamilies().get(1);
+        record.setDataBinary(((ByteString) data.get(0).getValue()).toByteArray());
```
Comment (hanahmily) [inline oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/deserializer/SegmentRecordDeserializer.java:50]: The module installer you don't create contains the metadata on how the index number maps to the tag. I would like to conceal these numbers under the installer, which is easier to maintain the codes.

**Review by hanahmily — COMMENTED** [review]
A quick view of the codes

File: oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageConfig.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    private String group = "default";
+    /**
+     * The maximum size of write entities in a single batch write call.
+     */
+    private int maxBulkSize = 5000;
+    /**
+     * Period of flush interval. In the timeunit of seconds.
+     */
+    private int flushInterval = 15;
```
[the remark this answers is not in the document]
Comment (KirannKaur) [reply oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageConfig.java:41]: @sonatype-lift help

**Comment by wu-sheng** [comment]
Before you do more integration, the metadata query DAO has re-implemented due to metadata-v2 query API accepted for SkyWalking v9.

Please sync the `banyandan-integration` branch from the master.

**Pull request closed without merging**